orangecontrib/text/widgets/owconcordance.py

```diff
@@ -0,0 +1,265 @@
+from orangecontrib.text.preprocess import Preprocessor, WordPunctTokenizer
+
+
+LastDocumentRole = next(gui.OrangeUserRole)
+
+
+class HorizontalGridDelegate(QStyledItemDelegate):
+    """Class for drawing horizontal line between documents."""
+    def paint(self, painter, option, index):
+        if index.data(LastDocumentRole):
+            painter.save()
+            painter.setPen(QColor(212, 212, 212))
+            painter.drawLine(option.rect.bottomLeft(), option.rect.bottomRight())
+            painter.restore()
+        if index.column() == 0:
+            option.textElideMode = Qt.ElideLeft
+        elif index.column() == 2:
+            option.textElideMode = Qt.ElideRight
+        QStyledItemDelegate.paint(self, painter, option, index)
+
+
+class DocumentSelectionModel(QItemSelectionModel):
+    """Sets selection for QTableView. Creates a set of selected documents."""
+
+    def select(self, selection, flags):
+        # which rows have been selected
+        indexes = selection.indexes() if isinstance(selection, QItemSelection) else [selection]
+        # indexes[0].row() == -1 indicates clicking outside of the table
+        if len(indexes) == 1 and indexes[0].row() == -1:
+            self.clear()
+            return
+        word_index = self.model().word_index
+        selected_docs = {word_index[index.row()][0] for index in indexes}
+        selected_rows = [row_index for row_index, (doc_index, _) in enumerate(word_index)
+                         if doc_index in selected_docs]
+        selection = QItemSelection()
+        # select all rows belonging to the selected document
+        for row in selected_rows:
+            index = self.model().index(row, 0)
+            selection.select(index, index)
+        super().select(selection, flags)
+
+
+class ConcordanceModel(QAbstractTableModel):
+    "A model for constructing concordances from text."
+
+    def __init__(self):
+        QAbstractTableModel.__init__(self)
+        self.word = None
+        self.data = None
+        self.indices = None
+        self.word_index = None
+        self.width = 8
+
+    def set_word(self, word):
+        self.modelAboutToBeReset.emit()
+        self.word = word
+        if self.indices is None:
+            self.word_index = None
+        else:
+            self.word_index = [(doc_idx, offset) for doc_idx, doc in enumerate(self.indices)
+                               for offset in doc.offsets(self.word)]
+        self.modelReset.emit()
+
+    def set_data(self, data):
+        self.modelAboutToBeReset.emit()
+        self.data = data
+        self._compute_indices()
+        self.modelReset.emit()
+
+    def set_width(self, width):
+        self.modelAboutToBeReset.emit()
+        self.width = width
+        self.modelReset.emit()
+
+    def flags(self, _):
+        return Qt.ItemIsEnabled | Qt.ItemIsSelectable
+
+    def rowCount(self, parent=None, *args, **kwargs):
+        return len(self.word_index) if self.word_index is not None else 0
```

Comment:
From what I read, parent.isValid() always returns False, since it [works only for trees not tables](http://stackoverflow.com/questions/38998467/how-to-construct-a-qmodelindex-with-valid-parent). Is this correct?